src/main/java/io/github/moulberry/notenoughupdates/miscfeatures/FairySouls.java

```diff
@@ -21,44 +21,273 @@
+	private boolean trackSouls;
+	private boolean showSouls;
+	private HashMap<String, HashMap<String, Set<Integer>>> allProfilesFoundSouls = new HashMap<>();
+	private List<BlockPos> allSoulsInCurrentLocation;
+	private Set<Integer> foundSoulsInLocation;
+	private TreeMap<Double, BlockPos> missingSoulsDistanceSqMap;
+	private List<BlockPos> closestMissingSouls;
+	private String currentLocation;
+	private BlockPos lastPlayerPos;
+
+	public static FairySouls getInstance() {
+		if (instance == null) {
+			instance = new FairySouls();
+		}
+		return instance;
+	}
+
+	@SubscribeEvent
+	public void onWorldLoad(WorldEvent.Load event) {
+		currentLocation = null;
+		trackSouls = NotEnoughUpdates.INSTANCE.config.misc.trackFairySouls;
+		showSouls = NotEnoughUpdates.INSTANCE.config.misc.fariySoul;
+	}
+
+	public void initializeLocation() {
+		if (!trackSouls || currentLocation == null) {
+			return;
+		}
+
+		foundSoulsInLocation = null;
+		closestMissingSouls = new ArrayList<>();
+		missingSoulsDistanceSqMap = new TreeMap<>();
+		lastPlayerPos = BlockPos.ORIGIN;
+
+		allSoulsInCurrentLocation = loadLocationFairySoulsFromConfig(currentLocation);
+		if (allSoulsInCurrentLocation == null) {
+			return;
+		}
+
+		foundSoulsInLocation = getFoundSoulsForProfile()
+			.computeIfAbsent(currentLocation, k -> new HashSet<>());
+		refreshMissingSoulInfo(true);
+	}
+
+	private void refreshMissingSoulInfo(boolean force) {
+		if (allSoulsInCurrentLocation == null) return;
+
+		BlockPos currentPlayerPos = Minecraft.getMinecraft().thePlayer.getPosition();
+		if (lastPlayerPos.equals(currentPlayerPos) && !force) {
+			return;
+		}
+		lastPlayerPos = currentPlayerPos;
+
+		missingSoulsDistanceSqMap.clear();
+		for (int i = 0; i < allSoulsInCurrentLocation.size(); i++) {
+			if (foundSoulsInLocation.contains(i)) {
+				continue;
+			}
+			BlockPos pos = allSoulsInCurrentLocation.get(i);
+			double distSq = pos.distanceSq(lastPlayerPos);
+			missingSoulsDistanceSqMap.put(distSq, pos);
+		}
+		closestMissingSouls.clear();
+		if (missingSoulsDistanceSqMap.isEmpty()) {
+			return;
+		}
+
+		// rebuild the list of the closest ones
+		int maxSouls = 15;
+		int souls = 0;
+		for (BlockPos pos : missingSoulsDistanceSqMap.values()) {
+			closestMissingSouls.add(pos);
+			if (++souls >= maxSouls) break;
+		}
+	}
+
+	private int interpolateColors(int color1, int color2, double factor) {
+		int r1 = ((color1 >> 16) & 0xff);
+		int g1 = ((color1 >> 8) & 0xff);
+		int b1 = (color1 & 0xff);
+
+		int r2 = (color2 >> 16) & 0xff;
+		int g2 = (color2 >> 8) & 0xff;
+		int b2 = color2 & 0xff;
+
+		int r3 = r1 + (int)Math.round(factor * (r2-r1));
+		int g3 = g1 + (int)Math.round(factor * (g2-g1));
+		int b3 = b1 + (int)Math.round(factor * (b2-b1));
+
+		return (r3 & 0xff) << 16 |
+			(g3 & 0xff) << 8 |
+			(b3 & 0xff);	}
+
+	private double normalize(double value, double min, double max) {
+		return ((value - min) / (max - min));
+	}
```

Comment:
Those two methods feel like they could belong in a utility class. (on the other hand our `Utils` class is extremely bloated and should probably be spread out over multiple files)

Reply:
Yeah, I actually looked for a good utility class to put these in and wasn't up for doing the cleanup so I left them here rather than adding them to the mess.